### PR TITLE
Search user

### PR DESCRIPTION
**Pull request by rohit-kumar6**

### CODE REVIEW

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 1]
```diff
@@ -1,16 +1,62 @@
...
-        return "";
+    @Autowired
+    private UserRepository userRepository;
+    @Autowired
+    private AccountRepository accountRepository;
+    @PersistenceContext
+    private EntityManager em;
+
+    private List<User> getUserById(List o){
+        List<User> list_user = new ArrayList<>();
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Camel case.

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 2]
```diff
@@ -1,16 +1,62 @@
...
+        for(Object id: o){
+            User user = userRepository.findById(((BigInteger) id).longValue()).orElseThrow();
+            list_user.add(user);
+        }
+        return list_user;
+    }
+
+    private List getResultBasedOnUsername(EntityManager em, String query){
+        Query q =  em.createNativeQuery("select id from accounts where username LIKE ? ");
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Add TODO for improvement.

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 3]
```diff
@@ -1,16 +1,62 @@
...
 
+    private List getResultBasedOnName(EntityManager em, String query){
+        Query q =  em.createNativeQuery("select id from users where name LIKE ? ");
+        q.setParameter(1, "%"+query+"%");
+        return q.getResultList();
+    }
+
+    @GetMapping("/search/{q}")
+    public List getSearchResultViaAjax(@PathVariable(value = "q") String query) throws EmptyQuery {
+        if(query.length()==0)
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Trim query.

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 4]
```diff
@@ -1,16 +1,62 @@
...
+    private List<User> getUserById(List o){
+        List<User> list_user = new ArrayList<>();
+        for(Object id: o){
+            User user = userRepository.findById(((BigInteger) id).longValue()).orElseThrow();
+            list_user.add(user);
+        }
+        return list_user;
+    }
+
+    private List getResultBasedOnUsername(EntityManager em, String query){
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Change name of function.
```suggestion
    private List getUserIdsByUsername(EntityManager em, String query){
```

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 5]
```diff
@@ -1,16 +1,62 @@
...
-    public String getSearchResultViaAjax() {
-        return "";
+    @Autowired
+    private UserRepository userRepository;
+    @Autowired
+    private AccountRepository accountRepository;
+    @PersistenceContext
+    private EntityManager em;
+
+    private List<User> getUserById(List o){
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Change names.
```suggestion
    private List<User> getUsersByIds(List ids){
```

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 6]
```diff
@@ -1,16 +1,62 @@
...
 
-    @PostMapping("/search")
-    public String getSearchResultViaAjax() {
-        return "";
+    @Autowired
+    private UserRepository userRepository;
+    @Autowired
+    private AccountRepository accountRepository;
+    @PersistenceContext
+    private EntityManager em;
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Change name.

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java (hunk 3 above)
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Validations on query parameter that all characters are valid.

**Review by NitronR — CHANGES_REQUESTED** [review]

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 7]
```diff
@@ -1,16 +1,82 @@
...
+        for(Object id: ids){
+            User user = userRepository.findById(((BigInteger) id).longValue()).orElseThrow();
+            listUser.add(user);
+        }
+        return listUser;
+    }
+
+    private List getUserIdsByUsername(EntityManager entityManager, String query){
+        // TODO : Optimise query
+        Query q =  entityManager.createNativeQuery("select id from accounts where username LIKE ? ");
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Do `select *`
```suggestion
        Query q =  entityManager.createNativeQuery("select * from accounts where username LIKE ? ");
```

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 8]
```diff
@@ -1,16 +1,82 @@
...
+        List<User> listUser = new ArrayList<>();
+        for(Object id: ids){
+            User user = userRepository.findById(((BigInteger) id).longValue()).orElseThrow();
+            listUser.add(user);
+        }
+        return listUser;
+    }
+
+    private List getUserIdsByUsername(EntityManager entityManager, String query){
+        // TODO : Optimise query
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Update TODO
```suggestion
        // TODO : Replace with hibernate search
```

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 9]
```diff
@@ -1,16 +1,82 @@
...
+        Query q =  entityManager.createNativeQuery("select id from accounts where username LIKE ? ");
+        q.setParameter(1, "%"+query+"%");
+        return q.getResultList();
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Typecast to account list.
```suggestion
        return getUsersFromAccounts((List<Account>)(Object)q.getResultList());
```

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 10]
```diff
@@ -1,16 +1,82 @@
...
+    private EntityManager entityManager;
+
+    private List<User> getUsersByIds(List ids){
+        List<User> listUser = new ArrayList<>();
+        for(Object id: ids){
+            User user = userRepository.findById(((BigInteger) id).longValue()).orElseThrow();
+            listUser.add(user);
+        }
+        return listUser;
+    }
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java:40]: Function not required. Add this function instead:
```
private List<User> getUsersFromAccounts(List<Account> accounts){
    // write logic here
}
```

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 11]
```diff
@@ -1,16 +1,82 @@
...
+
+    private List getUserIdsByUsername(EntityManager entityManager, String query){
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Update function name and return type.
```suggestion
    private List<User> getUsersByUsername(EntityManager entityManager, String query){
```

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 12]
```diff
@@ -1,16 +1,82 @@
...
+        // TODO : Optimise query
+        Query q =  entityManager.createNativeQuery("select id from accounts where username LIKE ? ");
+        q.setParameter(1, "%"+query+"%");
+        return q.getResultList();
     }
 
+    private List getUserIdsByName(EntityManager entityManager, String query){
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Same changes here also.
```suggestion
    private List<User> getUsersByName(EntityManager entityManager, String query){
```

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 13]
```diff
@@ -1,16 +1,82 @@
...
+        // TODO : Optimise query
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Update TODO
```suggestion
        // TODO : Replace with hibernate search
```

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 14]
```diff
@@ -1,16 +1,82 @@
...
+        Query q =  entityManager.createNativeQuery("select id from users where name LIKE ? ");
+        q.setParameter(1, "%"+query+"%");
+        return q.getResultList();
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Typecast to list of users.
```suggestion
        return (List<User>)(Object) q.getResultList();
```

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java [hunk 15]
```diff
@@ -1,16 +1,82 @@
...
+    }
+
+    @GetMapping({"/search/{q}","/search/"})
+    public List getSearchResultViaAjax(@PathVariable(value = "q",required = false) String query){
+        try {
```
Comment (NitronR) [inline src/main/java/com/mymemor/mymemor/controller/SearchController.java:104]: Add TODO.
```suggestion
        // TODO add top N logic
        try {
```

**Review by NitronR — CHANGES_REQUESTED** [review]

File: src/main/java/com/mymemor/mymemor/controller/SearchController.java (hunk 7 above)
Comment (rohit-kumar6) [reply src/main/java/com/mymemor/mymemor/controller/SearchController.java]: Better we can use
"select user_id from accounts where username LIKE ? "

**Pull request merged**